### PR TITLE
[flutter_tools] Deferred components setup validator

### PR DESCRIPTION
**Pull request by GaryQian**
Split off from https://github.com/flutter/flutter/pull/63773

Part of series of PRs for https://github.com/flutter/flutter/issues/73345 and https://github.com/flutter/flutter/issues/57617

This PR contains the `DeferredComponentsSetupValidator` which implements various checks and tasks that are used to verify the developer has a properly configured app for deferred components.

Most checks will look for the existence or particular entries inside of the android directory and if the file is not satisfactory, it will output the correct/expected file in a temporary directory under `build/android_deferred_components_setup_files`. Due to regex based parsers being unreliable and difficult to maintain, I will be switching to a Xml packaged based system to generate new additions to the xml files.

Example usage:

```
final DeferredComponentSetupValidator validator = DeferredComponentSetupValidator(environment);

validator.checkBlah1(components);
validator.checkBlah2(loadingUnits);

validator.handleResults();

// ---------------------------------------------------------
// Calling the following instead of handleResults is also valid:
validator.displayResults();
validator.attemptToolExit();
```

The validator is used in an upcoming PR based off of https://github.com/flutter/flutter/pull/63773.

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:95]: At a high level, having a single method that calls other methods based on a list of enums set in the object's constructor is incredibly confusing. Some of the checks aren't really checks either, like clear temp dir.

I would start by making each of these tasks performed by a public method, which documents what it is doing and named appropriately whether it is performing some cleanup or verification. I would also leave out this enum setup, at least for now, and we can discuss that more when you're further along the integration side of things.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:98]: These sorts of comments should really be captured in dartdoc if they're non-trivial

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:95]: (That would also remove a lot of the confusion about arguments that are only required conditionally)

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:307]: In general RegExps should be hoisted into statics and extensively documented as to what they are intended to match.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:593]: Do you need a specific diff  format here, or is there a way to do this in Dart? In general, we need to be really careful about introducing new runtime dependencies.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:602]: This would need to be done before this feature is landed

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:605]: Prefer to inject globals into the constructor

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:630]: If you use printError that will end up with error/red text. I would not spend too much time specifically formatting the errors

File: packages/flutter_tools/test/general.shard/android/deferred_components_setup_validator_test.dart
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/android/deferred_components_setup_validator_test.dart:44]: This is a dangerous pattern because forgetting to clear might lead to future tests passing inadvertently

Comment (GaryQian) [reply packages/flutter_tools/test/general.shard/android/deferred_components_setup_validator_test.dart:44]: I'll use the setup method to avoid this.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (GaryQian) [reply packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:593]: I prefer unidiff so that I can recommend a `patch` command to quickly apply all diffs, but if that is not possible the diff can still be displayed in other formats for developer reference. Overall, this step is not necessary for the function of this feature, but it provides helpful quality of life improvements.

I am not aware of any fully-dart ways to do this. The diff commands are more complex than it seems. For unix, `diff` should always be available. On windows however, `fc` is less powerful and doesn't generate the diffs I want. I am exploring using `git diff` instead. Since flutter requires git, git diff seems to be a relatively safe dependency.

In the worst case, if the commands for some reason cannot be run successfully, the diff step can be skipped without any material impact on the results of this validation.

Comment (GaryQian) [reply packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:568]: I'm actually going to remove the diff task from this PR. I'll land a more fully developed version of it in the future. Meanwhile I'd like to get the rest of the PR moving along.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:23]: out of date comment.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:90]: Can you link to an example/docs on what structure this is looking for in settings.gradle?

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:100]: Isn't this a user-owned file? How can we be sure that it can be safely deleted, and that the parsing will work correctly on valid but differently formatted files?

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:136]: prefer to write the file all at once, since it is not particularly big and to prevent cases where the tool getting killed would leave a partial file.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:108]: This seems like a very fragile check. Can you document the structure that you are trying to parse here? Is there any guarantee that this logic will continue to be valid?

**Comment by jonahwilliams** [comment]
How much of the automation that this PR is doing (reading/writing strings.xml/settings.gradle) is strictly necessary for an initial version of this feature?

I'm concerned that there is actually no good way to parse a settings.gradle (that I am aware of) - writing to user owned files in general is a bit of a trap.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:100]: ahh, misread `settingsGradle` and `settingsGradleOutput`

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:195]: nit: declare this closer to where it is used

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:216]: consider a more descriptive comment, or remove

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:213]: The tool should correctly throw errors for permissions into tool exits, so you might only hit a fse if the file is missing.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:216]: actually reading through some more, I would really like a higher level comment describing what this code is doing exactly, and what the mapping is

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:338]: nit: `// Entries` is not a very useful comment

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:300]: Could you expand on this a little more, what string resources are required, what is the expected format, et cetera

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:365]: nit: hoist and document

Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:365]: Shouldn't the xml parser give you access to names/properties in a way that doesn't require looking at raw lines?

Comment (GaryQian) [reply packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:365]: The idea is that I want to maintain the original formatting and comments in the file, and the xml parser consumes all of this and collapses it into a format-less xmldocument/xmlelement. This manual parsing is able to maintain any developer formatting/comments.

**Comment by GaryQian** [comment]
https://gist.github.com/GaryQian/9786e287b3c00406fa3c12fa62a30d7b is a snapshot of the validator before switching away from the RegExp based system for future reference.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (GaryQian) [reply packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart]: Moved to xml package based system.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:8]: nit: Alphabetize.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:32]: nit: indent continued lines 2 more spaces.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:30]: Should this parameter be `@required`?

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:42]: nit: public member comments.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:101]: Should there be a check here that this file exists, or is it already known to exist at this point?

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:105]: See: `ErrorHandlingFileSystem.deleteIfExists`

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:161]: nit: Avoid line breaks after assignment operators: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer-a-maximum-line-length-of-80-characters

File: packages/flutter_tools/test/general.shard/android/deferred_components_setup_validator_test.dart
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/android/deferred_components_setup_validator_test.dart:49]: Can these be `testWithoutContext`?

File: packages/flutter_tools/test/general.shard/android/deferred_components_setup_validator_test.dart
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/android/deferred_components_setup_validator_test.dart:98]: If this yaml can be provided by users by hand, then there should be tests that the tool does an appropriate `throwToolExit` when it is malformed.

File: packages/flutter_tools/test/general.shard/android/deferred_components_setup_validator_test.dart
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/android/deferred_components_setup_validator_test.dart:238]: Same idea here. If `settings.gradle` is under user control, then it should not crash the tool (from the new code in this PR) if it is malformed, and there should be unit tests of that.

**Review by zanderso — COMMENTED** [review]
High level comment to think about what happens when input files are malformed, and what happens when file system operations fail.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (GaryQian) [reply packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:30]: This property can be null. If null, it will just use the default templates directory. I can document this.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (GaryQian) [reply packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:129]: Added an `_invalidFiles` system to report errors like this.

**Comment by GaryQian** [comment]
I added systems to track and report file i/o and validity checking. Since this is an optional validator, any errors of this sort are reported in the validation output and are treated as if the check had failed (triggers tool exit)

**Review by jonahwilliams — COMMENTED** [review]
This PR is approaching the point where I don't think I could feasibly review it anymore. If each of the validations is fairly stand-alone - could you stage the PR so that each of them is added in a separate commit.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:100]: These methods can be made fairly standalone, by having each method individually return its inputs/outputs - then letting the caller be responsible for combining them.

Comment (GaryQian) [reply packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart]: So I actually want the class to handle the check result internally. I don't really want the caller to be handling the results of the validation since there are specific behaviors that should happen based off of the results of each check, which is handled by `displayResults`. Letting the caller combine/handle the check results directly can miss a lot of nuance and result in incorrect recommendations to the developer.

These checks should not really be run independently of the overall validator class.

**Comment by GaryQian** [comment]
Restructured commits to have each check added as a separate commit.

File: packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart:787]: Maybe something like this to avoid a ternary.

`final Directory outputDir = alternateAndroidDir?.childDirectory(name) ?? componentDir `

**Comment by jonahwilliams** [comment]
I'm really concerned about the parsing of the gradle settings file. I would feel more comfortable if we had better parsing, or at least knew what some of the edge cases were?

**Comment by GaryQian** [comment]
Ok, I'll just remove it for this version then. Consider it gone.

**Review by jonahwilliams — APPROVED** [review]
LGTM (if gradle.settings parsing is removed for now)

**Pull request merged**
